File: server/src/queues.js
// Read/modify the shared queues.yaml with comment- and order-preserving round-trips
// (the `yaml` Document API), guarded by a CROSS-PROCESS lock that the Python prune also
// takes. Both writers (this Node editor + queue_builder.queues.prune) run in the same
// container but as separate processes, so the Python threading lock can't cover us — a
// mkdir-based advisory lock on `<queues.yaml>.lock` does (see queue_builder/queues.py).
import { promises as fs } from 'node:fs';
import { parseDocument, YAMLSeq, Scalar } from 'yaml';
import { QUEUES_PATH } from './config.js';

const LOCK_DIR = QUEUES_PATH + '.lock';
const LOCK_STALE_MS = 15000; // a holder older than this is presumed dead; steal the lock
const LOCK_WAIT_MS = 10000; // give up acquiring after this
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function acquireLock() {
  const deadline = Date.now() + LOCK_WAIT_MS;
  for (;;) {
    try {
      await fs.mkdir(LOCK_DIR);
      return;
    } catch (e) {
      if (e.code !== 'EEXIST') throw e;
      // Steal a stale lock (a crashed holder that never rmdir'd).
      try {
        const st = await fs.stat(LOCK_DIR);
        if (Date.now() - st.mtimeMs > LOCK_STALE_MS) {
          await fs.rmdir(LOCK_DIR).catch(() => {});
          continue;
        }
      } catch {
        /* lock vanished between mkdir and stat — retry */
      }
      if (Date.now() > deadline) throw new Error('timed out acquiring queues.yaml lock');
      await sleep(50);
    }
  }
}

async function releaseLock() {
  await fs.rmdir(LOCK_DIR).catch(() => {});
}

async function withLock(fn) {
  await acquireLock();
  try {
    return await fn();
  } finally {
    await releaseLock();
  }
}

// Stable identity for an entry — MUST match queue_builder.queues.entry_key so the two
// writers address the same lines. `value` is a plain-JS entry (scalar or {ratingKey,title}).
export function entryKey(value) {
  if (value && typeof value === 'object' && !Array.isArray(value)) {
    if (value.ratingKey != null) return `rk:${value.ratingKey}`;
    if (value.title) return `title:${String(value.title).trim()}`;
    return null;
  }
  const s = String(value).trim();
  if (/^\d+$/.test(s)) return `rk:${s}`;
  return s ? `title:${s}` : null;
}

// A finished entry is KEPT and tagged by the Python service as a `{title/ratingKey, done: true}`
// mapping (decision: keep+tag rather than auto-prune). A plain string, a bare ratingKey, or a
// mapping without `done:true` is NOT done. Handles both on-disk shapes so a legacy plain entry
// simply reads as not-done.
export function entryDone(value) {
  return Boolean(value && typeof value === 'object' && !Array.isArray(value) && value.done === true);
}

async function readDoc() {
  let text = '';
  try {
    text = await fs.readFile(QUEUES_PATH, 'utf8');
  } catch (e) {
    if (e.code !== 'ENOENT') throw e;
  }
  const doc = parseDocument(text);
  if (!doc.contents || typeof doc.get !== 'function') doc.contents = doc.createNode({});
  return doc;
}

function seqFor(doc, setName) {
  let seq = doc.get(setName);
  if (!(seq instanceof YAMLSeq)) {
    seq = new YAMLSeq();
    doc.set(setName, seq);
  }
  return seq;
}

// Match the Python/ruamel writer's style so the file doesn't churn as the two writers
// alternate: `indentSeq: false` puts block dashes at the key's indent (ruamel offset=0),
// `lineWidth: 0` disables wrapping so long titles/comments stay on one line.
const YAML_OUT = { indentSeq: false, lineWidth: 0 };

async function writeDoc(doc) {
  const text = doc.toString(YAML_OUT);
  const tmp = QUEUES_PATH + '.tmp';
  await fs.writeFile(tmp, text, 'utf8');
  try {
    await fs.rename(tmp, QUEUES_PATH); // atomic on the same filesystem
  } catch {
    // A single-file bind-mount rejects rename-over (EBUSY); fall back to in-place write.
    await fs.writeFile(QUEUES_PATH, text, 'utf8');
    await fs.rm(tmp, { force: true }).catch(() => {});
  }
}

// Ordered raw entries for one set: [{ key, value }]. `value` is plain JS (scalar or object).
export async function listSet(setName) {
  const doc = await readDoc();
  const seq = doc.get(setName);
  if (!(seq instanceof YAMLSeq)) return [];
  return seq.items
    .map((node) => {
      const value = node.toJSON();
      return { key: entryKey(value), value, done: entryDone(value) };
    })
    .filter((e) => e.key !== null);
}

// Remove EVERY done entry from a set's list (the "Remove all completed" button). Done entries
// are the ones the Python service kept + tagged after they finished; this is the ONLY path
// that drops them (never automatic). Returns the count removed.
export async function removeCompleted(setName) {
  return withLock(async () => {
    const doc = await readDoc();
    const seq = doc.get(setName);
    if (!(seq instanceof YAMLSeq)) return { removed: 0 };
    const before = seq.items.length;
    seq.items = seq.items.filter((n) => !entryDone(n.toJSON()));
    const removed = before - seq.items.length;
    if (removed) {
      if (seq.items.length === 0) seq.flow = true; // restore a compact `[]` when emptied
      await writeDoc(doc);
    }
    return { removed };
  });
}

// Add a new entry. `value` is a string (title), a number (ratingKey), or {ratingKey,title}.
// `position` is 'top' (default — top plays next) or 'bottom'. Set-name validity is the
// caller's (server.js) job — it checks against the live sets.yaml registry.
export async function addItem(setName, value, position = 'top') {
  return withLock(async () => {
    const doc = await readDoc();
    const seq = seqFor(doc, setName);
    const key = entryKey(value);
    if (!key) throw new Error('empty entry');
    if (seq.items.some((n) => entryKey(n.toJSON()) === key)) return { added: false, key };
    seq.flow = false; // a populated queue is always a block list, never `[ ... ]`
    // Force double-quoted title strings so a `:` (e.g. "Star Trek: ...") stays readable.
    let node;
    if (typeof value === 'string') {
      node = doc.createNode(value);
      if (node instanceof Scalar) node.type = Scalar.QUOTE_DOUBLE;
    } else {
      node = doc.createNode(value);
    }
    if (position === 'bottom') seq.items.push(node);
    else seq.items.unshift(node);
    await writeDoc(doc);
    return { added: true, key };
  });
}

export async function removeItem(setName, key) {
  return withLock(async () => {
    const doc = await readDoc();
    const seq = doc.get(setName);
    if (!(seq instanceof YAMLSeq)) return { removed: false };
    const before = seq.items.length;
    seq.items = seq.items.filter((n) => entryKey(n.toJSON()) !== key);
    if (seq.items.length === before) return { removed: false };
    if (seq.items.length === 0) seq.flow = true; // restore a compact `[]` when emptied
    await writeDoc(doc);
    return { removed: true };
  });
}

function applyOrder(seq, keys) {
  const rank = new Map(keys.map((k, i) => [k, i]));
  const withKeys = seq.items.map((n, i) => ({ n, k: entryKey(n.toJSON()), i }));
  withKeys.sort((a, b) => {
    const ra = rank.has(a.k) ? rank.get(a.k) : keys.length + a.i;
    const rb = rank.has(b.k) ? rank.get(b.k) : keys.length + b.i;
    return ra - rb;
  });
  seq.items = withKeys.map((x) => x.n);
}

// Move an entry from one set to another (cross-queue drag), placing it per `toKeys` (the
// target set's desired key order incl. the moved entry). Same-set → a plain reorder. The
// actual YAML node is relocated, so its formatting/inline comment travels with it. Atomic:
// both sets live in one document, mutated under a single lock + one write.
export async function moveItem(fromSet, toSet, key, toKeys) {
  if (fromSet === toSet) return reorder(toSet, toKeys);
  return withLock(async () => {
    const doc = await readDoc();
    const src = doc.get(fromSet);
    if (!(src instanceof YAMLSeq)) return { moved: false };
    const idx = src.items.findIndex((n) => entryKey(n.toJSON()) === key);
    if (idx < 0) return { moved: false };
    const [node] = src.items.splice(idx, 1);
    if (src.items.length === 0) src.flow = true; // source emptied → compact `[]`
    const dst = seqFor(doc, toSet);
    dst.flow = false;
    if (!dst.items.some((n) => entryKey(n.toJSON()) === key)) dst.items.push(node);
    applyOrder(dst, toKeys);
    await writeDoc(doc);
    return { moved: true };
  });
}

// Bulk-move many entries (possibly from several source sets) into `toSet`, appended in the
// given order. One lock + one write, so a multi-select move is atomic. Entries already in
// `toSet` are left in place. `items` = [{fromSet, key}].
export async function moveBulk(items, toSet) {
  return withLock(async () => {
    const doc = await readDoc();
    const dst = seqFor(doc, toSet);
    let moved = 0;
    for (const { fromSet, key } of items) {
      if (fromSet === toSet) continue;
      const src = doc.get(fromSet);
      if (!(src instanceof YAMLSeq)) continue;
      const i = src.items.findIndex((n) => entryKey(n.toJSON()) === key);
      if (i < 0) continue;
      const [node] = src.items.splice(i, 1);
      if (src.items.length === 0) src.flow = true;
      if (!dst.items.some((n) => entryKey(n.toJSON()) === key)) {
        dst.flow = false;
        dst.items.push(node);
        moved += 1;
      }
    }
    if (moved) await writeDoc(doc);
    return { moved };
  });
}

// Bulk-remove entries across sets. `items` = [{fromSet, key}]. One lock + one write.
export async function removeBulk(items) {
  return withLock(async () => {
    const doc = await readDoc();
    let removed = 0;
    for (const { fromSet, key } of items) {
      const src = doc.get(fromSet);
      if (!(src instanceof YAMLSeq)) continue;
      const before = src.items.length;
      src.items = src.items.filter((n) => entryKey(n.toJSON()) !== key);
      if (src.items.length < before) {
        removed += 1;
        if (src.items.length === 0) src.flow = true;
      }
    }
    if (removed) await writeDoc(doc);
    return { removed };
  });
}

// An entry as {identity, extras}: `identity` is the ratingKey/title that makes the entry
// addressable, `extras` is every OTHER field the file carries (episodes, start, done, a
// hand-written `collection:`, …). Rewrites keep the extras, so setting one override never
// silently drops another writer's field.
function splitEntry(cur) {
  if (cur && typeof cur === 'object' && !Array.isArray(cur)) {
    const { ratingKey = null, title = null, ...extras } = cur;
    return { ratingKey, title, extras };
  }
  const s = String(cur).trim();
  if (/^\d+$/.test(s)) return { ratingKey: s, title: null, extras: {} };
  return { ratingKey: null, title: String(cur), extras: {} };
}

// Rebuild an entry node from its identity + extras, collapsing to the plainest form (a bare
// title string / ratingKey scalar) when there are no extras left to carry.
function entryNode(doc, { ratingKey, title, extras }) {
  const keys = Object.keys(extras).filter((k) => extras[k] != null);
  if (!keys.length) {
    if (title != null && ratingKey == null) {
      const node = doc.createNode(title);
      // Force a double-quoted title so a `:` (e.g. "Star Trek: …") stays readable.
      if (node instanceof Scalar) node.type = Scalar.QUOTE_DOUBLE;
      return node;
    }
    if (ratingKey != null && title == null) {
      return doc.createNode(/^\d+$/.test(String(ratingKey)) ? Number(ratingKey) : ratingKey);
    }
  }
  const o = {};
  if (ratingKey != null) o.ratingKey = ratingKey;
  if (title != null) o.title = title;
  for (const k of keys) o[k] = extras[k];
  return doc.createNode(o);
}

// Replace one entry in a set, addressed by its stable key. `mutate({ratingKey,title,extras})`
// edits the split form in place; the node is rebuilt from the result.
async function rewriteEntry(setName, key, mutate) {
  return withLock(async () => {
    const doc = await readDoc();
    const seq = doc.get(setName);
    if (!(seq instanceof YAMLSeq)) return false;
    const idx = seq.items.findIndex((node) => entryKey(node.toJSON()) === key);
    if (idx < 0) return false;
    const split = splitEntry(seq.items[idx].toJSON());
    mutate(split);
    seq.items[idx] = entryNode(doc, split);
    seq.flow = false;
    await writeDoc(doc);
    return true;
  });
}

// Set a series entry's per-show `episodes` (episodes queued per play). Rewrites the entry as
// a mapping carrying its ratingKey/title identity + `episodes` (or drops the field / reverts
// to a bare scalar when set back to 1). Entry identity (key) is unchanged.
export async function setEpisodes(setName, key, episodes) {
  const n = Math.max(1, Math.min(parseInt(episodes, 10) || 1, 20));
  const ok = await rewriteEntry(setName, key, (e) => {
    if (n > 1) e.extras.episodes = n;
    else delete e.extras.episodes;
  });
  return ok ? { ok: true, episodes: n } : { ok: false };
}

// Normalize a manual START point off the wire. A SHOW start is {season, episode}; a
// COLLECTION start also names the member to begin at — `series` is that member's ratingKey
// (a hand-written YAML entry may name it by title instead), and season/episode are optional
// (a movie member has neither). Anything without a series AND without an episode is "no
// start" — i.e. back to automatic next-unwatched.
export function normalizeStart(start) {
  if (!start || typeof start !== 'object') return null;
  const hasSeries = start.series != null && String(start.series).trim() !== '';
  if (!hasSeries && start.episode == null) return null;
  const s = {};
  if (hasSeries) s.series = String(start.series).trim();
  if (start.episode != null) {
    s.season = Math.max(1, parseInt(start.season, 10) || 1);
    s.episode = Math.max(1, parseInt(start.episode, 10) || 1);
  }
  return s;
}

// Set (or clear) an entry's manual START floor — begin here, skipping earlier episodes (and,
// for a collection, earlier members) WITHOUT marking anything watched. Preserves the entry's
// identity and every other field it carries; pass start=null to revert to automatic.
export async function setStart(setName, key, start) {
  const s = normalizeStart(start);
  const ok = await rewriteEntry(setName, key, (e) => {
    if (s) e.extras.start = s;
    else delete e.extras.start;
  });
  return ok ? { ok: true, start: s } : { ok: false };
}

// Drop a deleted queue's whole YAML key (used by DELETE /api/sets/:id so a removed queue
// doesn't leave an orphaned list behind). Missing key = fine.
export async function deleteSetKey(setName) {
  return withLock(async () => {
    const doc = await readDoc();
    if (!doc.has(setName)) return { deleted: false };
    doc.delete(setName);
    await writeDoc(doc);
    return { deleted: true };
  });
}

// Reorder a set to match `keys` (entry keys, new order). Entries not named in `keys` keep
// their relative order at the end, so a concurrently-added line is never dropped.
export async function reorder(setName, keys) {
  return withLock(async () => {
    const doc = await readDoc();
    const seq = doc.get(setName);
    if (!(seq instanceof YAMLSeq)) return { reordered: false };
    applyOrder(seq, keys);
    await writeDoc(doc);
    return { reordered: true };
  });
}

File: queue_builder/queues.py
"""Curated-queue store: read + prune the hand-edited `queues.yaml` wishlist.

The six `source="queue"` sets (bob / bob_alice / family and their _anime
siblings) are NOT computed by rule — Bob writes an ordered list per set into
`queues.yaml` (top = next to play). The file lives in App-Configs, mounted RW at
`config.QUEUES_PATH`, and is re-read on every scan. A finished entry is KEPT and tagged
`done: true` (mark_done), never auto-pruned — anime has no "Season 2", so the finished
series stays visible as the anchor Bob adds the sequel next to (decision
2026-07-21-finished-entries-marked-done-not-pruned, superseding the auto-prune half of
2026-07-16-movie-queue-sets-yaml-wishlist.md). Removal is explicit only (`prune`, driven
by a Node endpoint / per-item delete).

Round-trip via ruamel.yaml so hand-written comments + ordering survive the prune. If
ruamel is unavailable (e.g. the read-only dry-run env without the image deps), we fall
back to PyYAML for READS; a prune-rewrite is skipped with a warning rather than
clobbering the file's formatting.

## Entry formats (decision 2026-07-20-queue-entries-are-title-strings.md)

An entry is any of:

  * a **human-readable title string** — the primary, hand-editable form:
      - `- Duel (1971)`                         title + year
      - `- 86 Eighty-Six (2021) [anidb-16172]`  title + year + a `[source-id]` guid hint
      - `- Cowboy Bebop`                         bare title
    resolved against the set's Plex section at scan time (see plex._resolve_title).
  * a bare **ratingKey** (`- 12345`) — precise; what the web UI / AI write.
  * a **mapping** carrying either/both (`- {ratingKey: 12345, title: "Megamind (2010)"}`)
    — a ratingKey with a human label. The `title:` value is itself parsed as a title
    string (so a mapping can also carry `(year)`/`[guid]` if it has no ratingKey).

A human never needs a ratingKey; titles resolve fresh each scan and their exact text is
never rewritten (an unresolved title is kept + flagged, not clobbered).
"""
import errno
import os
import re
import threading
import time
from contextlib import contextmanager

from . import config

# Intra-process guard: the prune runs in the Python service; this RLock serializes it
# against itself across the service's own threads.
_LOCK = threading.RLock()

# CROSS-process guard: the Node web editor (plex-channels-web) writes the SAME file from a
# SEPARATE process in the same container, so the RLock can't cover both. Both writers take
# a mkdir-based advisory lock on `<queues.yaml>.lock` (mkdir is atomic everywhere). The
# Node side implements the identical convention in web/src/queues.js — keep them in sync.
_LOCK_STALE_S = 15.0     # a holder older than this is presumed dead; steal it
_LOCK_WAIT_S = 10.0      # give up acquiring after this


@contextmanager
def _file_lock():
    lock_dir = config.QUEUES_PATH + ".lock"
    deadline = time.monotonic() + _LOCK_WAIT_S
    while True:
        try:
            os.mkdir(lock_dir)
            break
        except OSError as e:
            if e.errno != errno.EEXIST:
                raise
            try:                                 # steal a stale lock (crashed holder)
                if time.time() - os.stat(lock_dir).st_mtime > _LOCK_STALE_S:
                    try:
                        os.rmdir(lock_dir)
                    except OSError:
                        pass
                    continue
            except OSError:
                pass                             # lock vanished mid-check: retry mkdir
            if time.monotonic() > deadline:
                raise TimeoutError("timed out acquiring queues.yaml lock")
            time.sleep(0.05)
    try:
        yield
    finally:
        try:
            os.rmdir(lock_dir)
        except OSError:
            pass

SET_KEYS = ("bob", "bob_alice", "family",
            "bob_anime", "bob_alice_anime", "family_anime")

# `Title (YEAR) [source-id]` — year and guid hint both optional, stripped from the END so
# a title that itself contains parentheses/brackets earlier is left intact.
_YEAR_RE = re.compile(r"\s*\((\d{4})\)\s*$")
_GUID_RE = re.compile(r"\s*\[([^\]]+)\]\s*$")
# `Collection: <name>` — a title string that plays a whole Plex Collection in order
# (decision 2026-07-21-collections-as-ordered-entries). Case-insensitive prefix.
_COLLECTION_RE = re.compile(r"^\s*collection:\s*(.+)$", re.IGNORECASE)


def parse_title_string(text):
    """Split a title string into (title, year|None, guid_hint|None).

    Peels an optional trailing `[source-id]` guid hint, then an optional trailing
    `(YEAR)`, leaving the bare title. Order-tolerant because both are matched at the
    end and stripped in turn.
    """
    s = str(text).strip()
    guid = None
    m = _GUID_RE.search(s)
    if m:
        guid = m.group(1).strip()
        s = s[:m.start()].rstrip()
    year = None
    m = _YEAR_RE.search(s)
    if m:
        year = int(m.group(1))
        s = s[:m.start()].rstrip()
    return s.strip(), year, guid


def _is_rating_key(value):
    """True if `value` is (or stringifies to) a bare numeric ratingKey."""
    if isinstance(value, bool):
        return False
    if isinstance(value, int):
        return True
    return isinstance(value, str) and value.strip().isdigit()


def entry_key(entry):
    """Stable identity for a queue entry, used to match file lines during prune.

    A ratingKey (scalar or mapping) keys as `rk:<n>`; a title (scalar string or a
    mapping's `title:`) keys as `title:<verbatim text>`. So a `- 12345` and a
    `{ratingKey: 12345}` collide (same entry), and a hand-typed title line keys on its
    exact text — which is what prune compares, so the human's line is never rewritten.
    """
    if isinstance(entry, dict):
        rk = entry.get("ratingKey")
        if rk is not None:
            return f"rk:{rk}"
        coll = entry.get("collection")
        if coll:
            # A `{collection: X}` mapping keys identically to a `Collection: X` string, so
            # the two spellings collide (same entry) and mark-done can round-trip either.
            return f"title:Collection: {str(coll).strip()}"
        title = entry.get("title")
        return f"title:{str(title).strip()}" if title else None
    if _is_rating_key(entry):
        return f"rk:{str(entry).strip()}"
    text = str(entry).strip()
    return f"title:{text}" if text else None


def _describe(entry):
    """Normalize a raw queue entry into a resolution descriptor.

    Returns {key, ratingKey|None, title|None, year|None, guid|None, collection|None,
    episodes|None, done, raw}. `title` is the PARSED bare title (year/guid peeled into
    their own fields); `collection` is the name from a `Collection: <name>` string or a
    `{collection: <name>}` mapping (plays that whole Plex Collection in order); `done` is
    True for a finished-but-kept entry (marked, not pruned — see mark_done); `raw` is the
    original file value (unused by resolution, handy for debugging).
    """
    if isinstance(entry, dict):
        rk = entry.get("ratingKey")
        coll = entry.get("collection")
        title, year, guid = (None, None, None)
        if entry.get("title"):
            title, year, guid = parse_title_string(entry["title"])
        # A `Collection: X` string may also arrive inside a mapping's `title:`; peel it out.
        if coll is None and title:
            cm = _COLLECTION_RE.match(title)
            if cm:
                coll = cm.group(1).strip()
        return {"key": entry_key(entry),
                "ratingKey": None if rk is None else str(rk),
                "title": title or None, "year": year, "guid": guid,
                "collection": str(coll).strip() if coll else None,
                "episodes": entry.get("episodes"),
                # Manual START floor {season, episode}: begin the show here, skipping earlier
                # episodes WITHOUT marking them watched (resume-from-Crunchyroll / skip-a-saga).
                "start": entry.get("start"),
                "done": bool(entry.get("done")), "raw": entry}
    if _is_rating_key(entry):
        return {"key": entry_key(entry), "ratingKey": str(entry).strip(),
                "title": None, "year": None, "guid": None, "collection": None,
                "episodes": None, "done": False, "raw": entry}
    title, year, guid = parse_title_string(entry)
    cm = _COLLECTION_RE.match(title)
    coll = cm.group(1).strip() if cm else None
    return {"key": entry_key(entry), "ratingKey": None,
            "title": title or None, "year": year, "guid": guid,
            "collection": coll, "episodes": None, "done": False, "raw": entry}


def _ruamel():
    """Return a configured ruamel YAML round-trip handler, or None if unavailable."""
    try:
        from ruamel.yaml import YAML
    except Exception:  # noqa: BLE001 — dry-run env may lack ruamel; caller falls back
        return None
    y = YAML()
    y.preserve_quotes = True
    y.indent(mapping=2, sequence=2, offset=0)
    return y


def _load_raw():
    """Parse queues.yaml. Returns (data, handler) — data is the ruamel/py object (mutable
    for ruamel), handler is the ruamel YAML or None when PyYAML-read fallback was used."""
    path = config.QUEUES_PATH
    if not os.path.exists(path):
        return {}, None
    y = _ruamel()
    with open(path, "r", encoding="utf-8") as f:
        if y is not None:
            return (y.load(f) or {}), y
        import yaml  # PyYAML read fallback (loses comments on any later rewrite)
        return (yaml.safe_load(f) or {}), None


def entries(set_name):
    """Ordered list of resolution descriptors for a set, [] if the set/file is empty.

    Each descriptor (see _describe) carries a stable `key`, plus a ratingKey OR a parsed
    (title, year, guid) for title-string resolution. Blank/keyless entries are dropped.
    """
    with _LOCK:
        data, _ = _load_raw()
    seq = (data or {}).get(set_name) or []
    out = []
    for e in seq:
        desc = _describe(e)
        if desc["key"] is not None:
            out.append(desc)
    return out


def prune(set_name, keep_keys):
    """Rewrite `set_name`'s list to only entries whose `entry_key` is in `keep_keys`.

    Re-reads off disk first (an SMB edit may have landed), filters that set's sequence to
    the kept entries (in the file's own order, verbatim text preserved), and atomically
    os.replace()s the result so a reader never sees a half-written file. No-op (returns
    False) if ruamel is missing or the file doesn't exist — we never clobber formatting
    with a lesser writer.
    """
    keep = set(keep_keys)
    with _LOCK, _file_lock():
        y = _ruamel()
        path = config.QUEUES_PATH
        if y is None or not os.path.exists(path):
            return False
        with open(path, "r", encoding="utf-8") as f:
            data = y.load(f) or {}
        seq = data.get(set_name)
        if seq is None:
            return False
        # Delete finished entries in place (high→low index) so remaining items keep their
        # comments/anchors; ruamel handles CommentedSeq item deletion.
        for i in range(len(seq) - 1, -1, -1):
            if entry_key(seq[i]) not in keep:
                del seq[i]
        tmp = path + ".tmp"
        with open(tmp, "w", encoding="utf-8") as f:
            y.dump(data, f)
        try:
            os.replace(tmp, path)                # atomic: a reader never sees a half file
        except OSError:
            # os.replace can't rename OVER a single-file bind-mount (EBUSY) or across devices.
            # The deploy mounts the DIRECTORY (so this path shouldn't hit), but fall back to a
            # direct in-place rewrite so a single-file mount still prunes. Non-atomic, but the
            # lock + re-read-before-write already bound the risk to one lost line.
            with open(path, "w", encoding="utf-8") as f:
                y.dump(data, f)
            try:
                os.remove(tmp)
            except OSError:
                pass
        return True


def mark_done(set_name, keep_keys):
    """Tag the given entries **done** in place — kept in the file, excluded from play.

    This replaces the auto-prune of finished entries (decision
    2026-07-21-finished-entries-marked-done-not-pruned): anime has no "Season 2", so a
    finished series must stay visible as the anchor Bob adds the sequel next to. Removal
    is now explicit (a Node endpoint / per-item delete), never automatic.

    Mirrors prune's round-trip discipline (re-read off disk, transform, atomic rewrite). A
    scalar entry is converted to a mapping so it can carry the flag — a ratingKey scalar ->
    `{ratingKey: <n>, done: true}`, a title/`Collection:` scalar -> `{title: <text>, done:
    true}`; a mapping simply gains `done: true`, preserving its other fields/comments. Match
    is by `entry_key`, so the human's own text is never rewritten (only wrapped). No-op
    (False) if ruamel is missing, the file/set is absent, or nothing changed — we never
    clobber formatting with a lesser writer.
    """
    want = set(keep_keys)
    if not want:
        return False
    with _LOCK, _file_lock():
        y = _ruamel()
        path = config.QUEUES_PATH
        if y is None or not os.path.exists(path):
            return False
        from ruamel.yaml.comments import CommentedMap
        with open(path, "r", encoding="utf-8") as f:
            data = y.load(f) or {}
        seq = data.get(set_name)
        if seq is None:
            return False
        changed = False
        for i in range(len(seq)):
            item = seq[i]
            if entry_key(item) not in want:
                continue
            if isinstance(item, dict):
                if not item.get("done"):
                    item["done"] = True
                    changed = True
            else:
                m = CommentedMap()
                if _is_rating_key(item):
                    m["ratingKey"] = str(item).strip()
                else:
                    m["title"] = str(item).strip()
                m["done"] = True
                seq[i] = m
                changed = True
        if not changed:
            return False
        tmp = path + ".tmp"
        with open(tmp, "w", encoding="utf-8") as f:
            y.dump(data, f)
        try:
            os.replace(tmp, path)
        except OSError:
            with open(path, "w", encoding="utf-8") as f:
                y.dump(data, f)
            try:
                os.remove(tmp)
            except OSError:
                pass
        return True
